app/controllers/users_controller.rb

```diff
@@ -72,46 +72,81 @@ def destroy
+    # this hashmap is for keeping track of which column corresponds to an event id,
+    # because we have an unsorted bulk manual query that needs to be sorted for the table
+    event_id_map = {}
+    i = 0
+    events.each do |event|
+      row0.push(event.title)
+      marginal_event_attendance.push(0)
+      event_id_map[event.id] = i
+      i += 1
+    end
+
+    # for keeping track of which row corresponds to a user id
+    user_id_map = {}
+    i = 0
+    users.each do |user|
+      user_id_map[user.id] = i
+      i += 1
+    end
+
+    table = Array.new(users.length) { Array.new(events.length) { false } }
+
+    # set cells based on query
+    query.each do |pair|
+      user_indx = user_id_map[pair['customer_id']]
+      event_indx = event_id_map[pair['event_id']]
+      table[user_indx][event_indx] = true
+    end
+
+    # fill out second row with user attribute headers and event attendance marginals
+    row1 = ['Last name', 'First name', 'Email', '']
+    corner_marginal = 0
+    (0...events.length).each do |event_i|
+      column_marginal = 0
+      (0...users.length).each do |user_i|
+        column_marginal += 1 if table[user_i][event_i]
+      end
+      corner_marginal += column_marginal
+      row1.push(column_marginal)
+    end
+    row1[3] = "marginals total: #{corner_marginal}"
 
     # create csv string
     csv_string = CSV.generate do |csv|
-      # set the header row first
-      header = ['Last name', 'First name', 'total attendances']
-
-      # push a column for every meeting
-      event_index = header.length
-      # this hashmap is for keeping track of which column corresponds to an event id,
-      # because we have a manual query that would be clumsy to sort
-      id_map = {}
-      events.each do |event|
-        header.push(event.title)
-        id_map[event.id] = event_index
-        event_index += 1
-      end
-      csv << header
+      csv << row0
+      csv << row1
 
       # fill out table rows according to every user
+      i = 0
       users.each do |user|
-        row = [user.last_name, user.first_name, '']
-        # start the row empty
-        (0..id_map.size).each do
-          row.push('')
-        end
-        # manual query instead of needing to have an existence query for every combination of
-        # customer and event
-        query = ActiveRecord::Base.connection.execute(
-          "SELECT customers_events.event_id FROM customers_events WHERE customers_events.customer_id = #{user.id};"
-        )
-        num_attended = 0
-        query.each_row do |id|
-          # modify the corresponding cell to have a '1'
-          row[id_map[id.first]] = '1'
-          num_attended += 1
+        row_marginal = 0
+        row = [user.last_name, user.first_name, user.email, '']
+        table_row = table[i]
+        (0...events.length).each do |j|
+          if table_row[j]
+            row_marginal += 1
+            row.push('1')
+          else
+            row.push('')
```

Comment:
This is the only line that isn't covered by our test suite, which brings our coverage to 99.91%